File: backend/app/schemas/software.py
from pydantic import BaseModel, Field


class SoftwareDependencyIn(BaseModel):
    kind: str = "external"  # internal / external / runtime
    name: str = Field(default="", max_length=128)
    target_component_id: str | None = None
    constraint: str = Field(default="", max_length=64)


class SoftwareVersionCreate(BaseModel):
    version: str = Field(min_length=1, max_length=64)
    note: str = Field(default="", max_length=4000)
    artifact_type: str = Field(default="none", pattern="^(file|url|none)$")
    file_name: str = Field(default="", max_length=255)
    file_size: float = 0.0
    download_url: str = Field(default="", max_length=4000)
    published_at: str | None = None
    dependencies: list[SoftwareDependencyIn] = []


class SoftwareVersionUpdate(BaseModel):
    version: str | None = Field(default=None, min_length=1, max_length=64)
    note: str | None = Field(default=None, max_length=4000)
    artifact_type: str | None = Field(default=None, pattern="^(file|url|none)$")
    file_name: str | None = Field(default=None, max_length=255)
    file_size: float | None = None
    download_url: str | None = Field(default=None, max_length=4000)
    published_at: str | None = None
    dependencies: list[SoftwareDependencyIn] | None = None


class SoftwareComponentCreate(BaseModel):
    name: str = Field(min_length=1, max_length=128)
    identifier: str = Field(min_length=1, max_length=128)
    category: str = Field(default="other", max_length=32)
    icon_key: str = Field(default="", max_length=16)
    description: str = Field(default="", max_length=2000)


class SoftwareComponentUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=128)
    identifier: str | None = Field(default=None, min_length=1, max_length=128)
    category: str | None = Field(default=None, max_length=32)
    icon_key: str | None = Field(default=None, max_length=16)
    description: str | None = Field(default=None, max_length=2000)

File: frontend/src/api/software.ts
import api from "./client";

export interface DependencyData {
  id: string;
  kind: "internal" | "external" | "runtime";
  name: string;
  target_component_id: string | null;
  constraint: string;
}

export interface VersionData {
  id: string;
  component_id: string;
  version: string;
  note: string;
  artifact_type: "file" | "url" | "none";
  file_name: string;
  file_size: number;
  download_url: string;
  stored_filename: string;
  published_at: string;
  created_at: string;
  updated_at: string;
  dependencies: DependencyData[];
}

export interface ComponentData {
  id: string;
  name: string;
  identifier: string;
  category: string;
  icon_key: string;
  description: string;
  created_at: string;
  updated_at: string;
  version_count: number;
  latest_version: VersionData | null;
  dependency_count: number;
  internal_dependency_count: number;
  versions?: VersionData[] | null;
}

export interface SoftwareStats {
  component_count: number;
  version_count: number;
  artifact_count: number;
  dependency_count: number;
}

export interface DependencyInput {
  kind: string;
  name: string;
  target_component_id?: string | null;
  constraint: string;
}

export interface VersionInput {
  version: string;
  note?: string;
  artifact_type?: "file" | "url" | "none";
  file_name?: string;
  file_size?: number;
  download_url?: string;
  published_at?: string | null;
  dependencies?: DependencyInput[];
}

export interface ComponentInput {
  name: string;
  identifier: string;
  category?: string;
  icon_key?: string;
  description?: string;
}

export async function listComponents(
  params: { category?: string; sort?: string } = {}
): Promise<ComponentData[]> {
  const res = await api.get("/software", { params });
  return res.data;
}

export async function getStats(): Promise<SoftwareStats> {
  const res = await api.get("/software/stats");
  return res.data;
}

export async function getComponent(id: string): Promise<ComponentData> {
  const res = await api.get(`/software/${id}`);
  return res.data;
}

export async function createComponent(data: ComponentInput): Promise<ComponentData> {
  const res = await api.post("/software", data);
  return res.data;
}

export async function updateComponent(
  id: string,
  data: Partial<ComponentInput>
): Promise<ComponentData> {
  const res = await api.put(`/software/${id}`, data);
  return res.data;
}

export async function deleteComponent(id: string): Promise<void> {
  await api.delete(`/software/${id}`);
}

export async function createVersion(
  componentId: string,
  data: VersionInput
): Promise<VersionData> {
  const res = await api.post(`/software/${componentId}/versions`, data);
  return res.data;
}

export async function getVersion(id: string): Promise<VersionData> {
  const res = await api.get(`/software/versions/${id}`);
  return res.data;
}

export async function updateVersion(
  id: string,
  data: Partial<VersionInput>
): Promise<VersionData> {
  const res = await api.put(`/software/versions/${id}`, data);
  return res.data;
}

export async function deleteVersion(id: string): Promise<void> {
  await api.delete(`/software/versions/${id}`);
}

export async function uploadArtifact(
  versionId: string,
  file: File
): Promise<{ filename: string; original_name: string; size_mb: number; url: string }> {
  const form = new FormData();
  form.append("file", file);
  const res = await api.post(`/software/versions/${versionId}/upload`, form);
  return res.data;
}

/** Human-friendly artifact download URL on the SPA origin. */
export function artifactUrl(storedFilename: string): string {
  return `/api/v1/software/files/${storedFilename}`;
}
